consensus/event_loop.go

```diff
@@ -521,6 +550,23 @@ func (fc *FastHotStuffEventLoop) hasTimedOutForView(publicKey *bls.PublicKey, vi
 	return ok
 }
 
+func (fc *FastHotStuffEventLoop) getBlockAndValidatorSetByHash(blockHash BlockHash) (
+	bool, Block, []Validator, map[string]Validator,
+) {
+	// A linear search here is fine. The safeBlocks slice is expected to be extremely small as it represents the
+	// number of uncommitted blocks in the blockchain. During steady stake, it will have a size of 3 blocks
+	// (one committed, two uncommitted). In the worse case, where the network has an unlucky series of
+	// timeout -> block -> timeout -> block,... it can still be expected to have < 10 blocks.
+	blockHashValue := blockHash.GetValue()
+	for _, block := range fc.safeBlocks {
+		if block.block.GetBlockHash().GetValue() == blockHashValue {
```

Comment:
typically we like to use `bytes.Equal` to compare byte slices, but since these are `[32]byte` types we can't do that. Maybe we define a function to help with this. I can't quite remember if there was reason that doing a comparison with `==` would result in unexpected outcomes. Alternatively we add a `Eq` function to the interface definition of `BlockHash`, which the existing BlockHash struct already implements.

Reply:
You sure that using the `==` equality checks for doesn't work for arrays? Note, these aren't slices. Slices are reference types so it's ambiguous what `==` does underneath.

These are fixed length byte arrays, which means they're value types where the length (32) is part of the type definition. The `==` comparison performs a value comparison. I was confirmed to confirm this via https://go.dev/play/.

Reply:
Yeah you right. I think the issue was slices which as you say have ambiguous lengths.